src/pyfmi/fmi_base.pyx

```diff
@@ -0,0 +1,788 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+# Copyright (C) 2025Modelon AB
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation, version 3 of the License.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+# distutils: define_macros=NPY_NO_DEPRECATED_API=NPY_1_7_API_VERSION
+
+"""
+Module containing the FMI interface Python wrappers.
+"""
+
+"""
+For profiling:
+# cython: linetrace=True
+# cython: binding=True
+# distutils: define_macros=CYTHON_TRACE_NOGIL=1
+"""
+
+import os
+import logging
+import fnmatch
+import re
+from io import UnsupportedOperation
+cimport cython
+
+import functools
+import marshal
+
+import numpy as np
+cimport numpy as np
+
+cimport pyfmi.fmil_import as FMIL
+from pyfmi.exceptions import FMUException
+
+
+# from pyfmi.fmi_util cimport decode
+cpdef decode(x): # TODO; temporarily here due to circular dependency
+    if isinstance(x, bytes):
+        return x.decode(errors="replace")
+    else:
+        return x
+cpdef encode(x):
+    if isinstance(x, str):
+        return x.encode()
+    else:
+        return x
+
+
+int   = np.int32
+np.int = np.int32
+
+# FMI types
+FMI_DEFAULT_LOG_LEVEL = FMIL.jm_log_level_error
+
+cdef FMIL.fmi_version_enu_t import_and_get_version(FMIL.fmi_import_context_t* context, char* fmu_full_path, char* fmu_temp_dir, int allow_unzipped_fmu):
+    """ Invokes the necessary FMIL functions to retrieve FMI version while accounting
+        for the conditional 'allow_unzipped_fmu'.
+    """
+    if allow_unzipped_fmu:
+        return FMIL.fmi_import_get_fmi_version(context, NULL, fmu_temp_dir)
+    else:
+        return FMIL.fmi_import_get_fmi_version(context, fmu_full_path, fmu_temp_dir)
+
+# TODO: copied due to circular dependencies; replace by functools lru_cache?
```

Comment:
XXX: Could potentially be replaced, not sure how important the storing in binary from "marshal" is.